src/hotspot/share/jeandle/jeandleAbstractInterpreter.cpp

```diff
@@ -1909,10 +1921,29 @@ void JeandleAbstractInterpreter::null_check(llvm::Value* obj) {
   llvm::Value* if_null = _ir_builder.CreateICmp(llvm::CmpInst::ICMP_EQ,
                                                 obj,
                                                 llvm::ConstantPointerNull::get(llvm::cast<llvm::PointerType>(obj->getType())));
-  _ir_builder.CreateCondBr(if_null, null_check_fail, null_check_pass);
+  llvm::BranchInst* null_check_br = _ir_builder.CreateCondBr(if_null, null_check_fail, null_check_pass);
+
+  // Add make.implicit metadata, and the ImplicitNullChecksPass transforms explicit checks into implicit ones.
+  llvm::MDNode* make_implicit = llvm::MDNode::get(*_context, {});
+  null_check_br->setMetadata(llvm::LLVMContext::MD_make_implicit, make_implicit);
 
+  // TODO: workaround, we need an uncommon trap here.
   _ir_builder.SetInsertPoint(null_check_fail);
-  dispatch_exception_to_handler(llvm::ConstantPointerNull::get(llvm::cast<llvm::PointerType>(JeandleType::java2llvm(BasicType::T_OBJECT, *_context))));
+  llvm::FunctionCallee null_check_fail_callee = JeandleRuntimeRoutine::hotspot_SharedRuntime_throw_NullPointerException_callee(_module);
+  llvm::CallInst* current_thread = call_java_op("jeandle.current_thread", {});
+  llvm::CallInst* call_null_check_fail = _ir_builder.CreateCall(null_check_fail_callee, {current_thread});
+  llvm::Type* ret_type = _llvm_func->getReturnType();
+  if (ret_type->isVoidTy()) {
+    _ir_builder.CreateRetVoid();;
+  } else if (ret_type->isIntegerTy()) {
+    _ir_builder.CreateRet(llvm::ConstantInt::get(ret_type, 0));
+  } else if (ret_type->isFloatTy() || ret_type->isDoubleTy()) {
+    _ir_builder.CreateRet(llvm::ConstantFP::get(ret_type, 0.0));
+  } else if (ret_type->isPointerTy()) {
+    _ir_builder.CreateRet(llvm::ConstantPointerNull::get(llvm::cast<llvm::PointerType>(ret_type)));
+  } else {
+    ShouldNotReachHere();
+  }
```

Comment:
There is a similar logic in `JeandleAbstractInterpreter::throw_exception`, merge them as a function